### ISSUE TITLE
ARIA attribute "role" got stripped when passing custom ALLOWED_URI_REGEXP config that doesn't match "role"

### ISSUE THREAD

**Issue by karl19** [cure53/DOMPurify#554]
### Background & Context

ARIA attribute like "role" and "aria-*" are always considered whitelisted 

### Bug

"role" will be stripped when passing own ALLOWED_URI_REGEXP that does not match "role" like below, which sounds weird (yes i know)

```
 const source = DOMPurify.sanitize(dirty, {
  ALLOWED_URI_REGEXP:
       /^(https?|[^a-z]|[a-z+\-.]+(?:[^a-z+.\-:]))/i,
  });
```
#### Input
```
<div role="layout">bla bla bla</div>
```
#### Output by using above ALLOWED_URI_REGEXP as config
```
<div>bla bla bla</div>
```
#### Expected output
```
<div role="layout">bla bla bla</div>
```

### Suggestion & Question

quick fix would be adding "role" in https://github.com/cure53/DOMPurify/blob/main/src/regexp.js#L7? as it only match "aria-*" right now

IS_ALLOWED_URI in https://github.com/cure53/DOMPurify/blob/main/src/regexp.js#L8 seems to match everything, is it intended? it supposed not to have impact on the name of the attribute but the value?

**Comment by cure53** [comment]
Hmm, when I simply do this in latest DOMPurify, things work as expected :

```javascript
DOMPurify.sanitize('<div role="layout">bla bla bla</div>')
// "<div role=\"layout\">bla bla bla</div>"
```

Do you really need to add a "system wide" regex for all attribute values? 
What is the use case you want to cover here - why redefine the regex for all attribute values?

**Comment by karl19** [comment]
Yes it works without passing through customised ALLOWED_URI_REGEXP which we used to restrict to specific protocol per readme, but it seems it is not limited to the protocol but "system wide" attributes' value, if so it is probably a misuse of it, for our use case we then have to apply it in the hook when we want to validate the value of eg. `href` for anchor?

**Comment by cure53** [comment]
Exactly, it is system wide for the reason that we cannot be sure which JS framework uses which attribute to handle URLs, and we had XSS in the early days because of that.

Generally, DOMPurify should already make sure that only harmless URL schemes can be used (this hasn't been bypassed for a loooong time which is some form of metric :D). But it is possible to restrict that even further if needs be, via hook:

https://github.com/cure53/DOMPurify/blob/main/demos/hooks-link-proxy-demo.html

I think this might be a way of doing it in a less impactful way but still maintain the same level of security?

**Comment by cure53** [comment]
Or, another way of doing it:

https://github.com/cure53/DOMPurify/blob/main/demos/hooks-scheme-allowlist.html

**Comment by karl19** [comment]
Thanks @cure53, if so we will definitely change the way how we restrict on the protocol. But meantime I still think "role" should be made into either ARIA_ATTR or URI_SAFE_ATTRIBUTES...

**Comment by cure53** [comment]
I think you are right with `URI_SAFE_ATTRIBUTES`

Wanna spin up a PR? :)

**Comment by karl19** [comment]
I don't mind, but looks like I have no permission to do so...

![image](https://user-images.githubusercontent.com/481215/128432675-afb26277-e423-4610-91d7-678921ba0dfb.png)

**Comment by karl19** [comment]
btw, i got a question on conditionally whitelisting a tag, do i have to do the check and removal in hook?

**Comment by karl19** [comment]
nws, got the example :)

**Comment by bcaller** [comment]
I had a similar query but I'll just explain how I fixed it in case it helps someone.

When `ALLOWED_URI_REGEXP` is set (e.g. to limit protocol to only http(s)), the `OL` tag's `start` attribute is stripped from `<ol start="5">`.

According to the README, we can use `ALLOWED_URI_REGEXP` to

> allow specific protocols handlers in URL attributes via regex

So I used `/^(?:(?:https?):\/\/|\/)/i`.

However, this regex is being applied to most attributes not in https://github.com/cure53/DOMPurify/blob/926a8cd1755e14bd1523c4e034e4c01f361e9218/src/purify.js#L343-L358

I thought this was a bug but as mentioned above:

> we cannot be sure which JS framework uses which attribute to handle URLs, and we had XSS in the early days because of that

So it checks all unknown attributes to make sure they aren't URIs with unknown protocols.

Solution for me was just carefully modifying the regexp from the default

```js
/^(?:(?:(?:f|ht)tps?|mailto|tel|callto|sms|cid|xmpp):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i
```

to remove the unwanted protocols but still allowing attributes like `"layout"`, `"/xyz"` or `"5"`

```js
/^(?:https?:|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i
```
